### PR TITLE
Add subscribeEvents and subcriptionEvent

### PR DESCRIPTION
**Pull request by IronGauntlets**

### CODE REVIEW

File: rpc/subscriptions.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	defer func() {
+		h.callAndLogErr(filter.Close, "Error closing event filter in events subscription")
+	}()
+
+	if err = setEventFilterRange(filter, &BlockID{Number: from}, &BlockID{Number: to}, to); err != nil {
+		h.log.Warnw("Error setting event filter range", "err", err)
+		return
+	}
+
+	filteredEvents, cToken, err := filter.Events(nil, subscribeEventsChunkSize)
```
Comment (pnowosie) [inline rpc/subscriptions.go:123]: How about `nextToken` istead of `cToken`? - I'd like to make the name more descriptive.

File: rpc/subscriptions_test.go [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+		serverConn, clientConn := net.Pipe()
+		t.Cleanup(func() {
+			require.NoError(t, serverConn.Close())
+			require.NoError(t, clientConn.Close())
+		})
+
+		subCtx := context.WithValue(context.Background(), jsonrpc.ConnKey{}, &fakeConn{w: serverConn})
+
+		// Note the end of the window doesn't need to be tested because if requested block number is more than the
+		// head, a block not found error will be returned. This behaviour has been tested in various other test, and we
```
Comment (pnowosie) [inline rpc/subscriptions_test.go:119]:
```suggestion
		// head, a block not found error will be returned. This behaviour has been tested in various other tests, and we
```

File: rpc/subscriptions_test.go [hunk 3]
```diff
@@ -0,0 +1,338 @@
+package rpc
```
Comment (pnowosie) [inline rpc/subscriptions_test.go:1]: Usualy tests go to `rpc_test` pkg. Can you add a comment with a reason here?

**Review by pnowosie — COMMENTED** [review]
Some minors comments & ask for details

File: rpc/handlers.go [hunk 4]
```diff
@@ -65,12 +65,15 @@ var (
 	ErrUnsupportedTxVersion            = &jsonrpc.Error{Code: 61, Message: "the transaction version is not supported"}
 	ErrUnsupportedContractClassVersion = &jsonrpc.Error{Code: 62, Message: "the contract class version is not supported"}
 	ErrUnexpectedError                 = &jsonrpc.Error{Code: 63, Message: "An unexpected error occurred"}
+	ErrTooManyBlocksBack               = &jsonrpc.Error{Code: 68, Message: "Cannot go back more than 1024 blocks"}
```
Comment (kirugan) [inline rpc/handlers.go:68]: do you think it's worth to pass actual constant here i.e.:
```
Message: fmt.Sprintf("Cannot go back more than %d blocks", maxBlocksBack)
```

File: rpc/subscriptions.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+				case <-subscriptionCtx.Done():
+					return
+				case header := <-headerSub.Recv():
+
+					h.processEvents(subscriptionCtx, w, id, header.Number, header.Number, fromAddr, keys)
+				}
+			}
+		}()
+
+		h.processEvents(subscriptionCtx, w, id, requestedHeader.Number, headHeader.Number, fromAddr, keys)
```
Comment (kirugan) [inline rpc/subscriptions.go:99]: Could you elaborate what is difference between this call here and another one in a goroutine above

File: rpc/subscriptions.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+func (h *Handler) processEvents(ctx context.Context, w jsonrpc.Conn, id, from, to uint64, fromAddr *felt.Felt, keys [][]felt.Felt) {
+	filter, err := h.bcReader.EventFilter(fromAddr, keys)
+	if err != nil {
+		h.log.Warnw("Error creating event filter", "err", err)
+		return
+	}
+
+	defer func() {
+		h.callAndLogErr(filter.Close, "Error closing event filter in events subscription")
+	}()
```
Comment (kirugan) [inline rpc/subscriptions.go:116]: seems like lambda is redundant here:
```
defer h.callAndLogErr(filter.Close, "...")
```

Comment (IronGauntlets) [reply rpc/subscriptions.go]: Done

File: rpc/subscriptions.go (hunk 5 above)
Comment (IronGauntlets) [reply rpc/subscriptions.go:99]: The one in the goroutine checks for events from new blocks as it listens for the new heads subscription.

This one fetches the events from current and/or old blocks especially if the requested blockID requests events from block `curHeight-1024`

File: rpc/handlers.go (hunk 4 above)
Comment (IronGauntlets) [reply rpc/handlers.go]: Done

File: rpc/subscriptions_test.go (hunk 3 above)
Comment (IronGauntlets) [reply rpc/subscriptions_test.go:1]: Yes but we don't have to follow it if its is difficult to use components which make network/db operations. Therefore, I use `package rpc` for mocking which allows me to easily make particular scenarios on events possible.

File: rpc/subscriptions_test.go (hunk 2 above)
Comment (IronGauntlets) [reply rpc/subscriptions_test.go]: Done

File: rpc/subscriptions.go (hunk 1 above)
Comment (IronGauntlets) [reply rpc/subscriptions.go:121]: `c` is short for continuation, same convention is used in other event related functions.

File: rpc/subscriptions.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
+package rpc
```
[the remark this answers is not in the document]
Comment (IronGauntlets) [reply rpc/subscriptions.go:1]: I think we should move all the subscription related code to `subscriptions.go` and other prs should follow the same convention.

**Review by kirugan — APPROVED** [review]

**Review by pnowosie — APPROVED** [review]

**Pull request merged**